docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,259 @@
+
+8.  The existing alpha dynamic provisioning feature be phased out in the
+    next release.
+
+### Controller workflow for provisioning volumes
+
+1.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.
+
+    1. If the claim has non-empty `claim.Spec.Class`, only PVs with the same
+        `pv.Spec.Class` are considered.
+
+    2. If the claim has empty `claim.Spec.Class`, all existing PVs are
+        considered.
+
+    All "considered" volumes are evaluated and the
+    smallest matching volume is bound to the claim.
+
+2.  If no volume is found for the claim and `claim.Spec.Class` is not set or is
+    empty string dynamic provisioning is disabled.
+
+3.  If `claim.Spec.Class` is set the controller tries to find instance of StorageClass with this name.  If no
+    such StorageClass is found, the controller goes back to step 1. and
+    periodically retries finding a matching volume or storage class again until
+    a match is found. The claim is `Pending` during this period.
+
+4.  With StorageClass instance, the controller finds volume plugin specified by
+    StorageClass.ProvisionerType.
+
+5.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin`, which has a method called `NewProvisioner`
+    that returns a new provisioner.
+
+6.  The controller calls volume plugin `Provision` with ProvisionerParameters from the `StorageClass` configuration object. 
+
+7.  If `Provision` returns an error, the controller generates an event on the
+    claim and goes back to step 1., i.e. it will retry provisioning periodically
+
+8.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, fills its `Class` attribute with `claim.Spec.Class`
+    and makes it already bound to the claim
+
+  1.  If the create operation for the `api.PersistentVolume` fails, it is
+      retried
+
+  2.  If the create operation does not succeed in reasonable time, the
+      controller attempts to delete the provisioned volume and creates an event
+      on the claim
+
+Existing behavior is un-changed for claims that do not specify `claim.Spec.Class`.
+
+### `StorageClass` API
+
+A new API group should hold the API for storage classes, following the pattern
+of autoscaling, metrics, etc.  To allow for future storage-related APIs, we
+should call this new API group `storage` and incubate in extensions/v1beta1.
+
+Storage classes will be represented by an API object called `StorageClass`:
+
+```go
+package storage
+
+// StorageClass describes the parameters for a class of storage for
+// which PersistentVolumes can be dynamically provisioned.
+//
+// StorageClasses are non-namespaced; the name of the storage class
+// according to etcd is in ObjectMeta.Name.
+type StorageClass struct {
```

Comment:
To think about: is StorageClass updatable?  What does it mean if admin redefines "Purple" while volumes that claim to be purple exist?